lms/services/grouping.py

```diff
@@ -0,0 +1,68 @@
+from lms.models import CanvasSection, Grouping
+from lms.models._hashed_id import hashed_id
+
+
+class GroupingService:
+    def __init__(self, db, application_instance_service, course_service):
+        self._db = db
+        self._application_instance = application_instance_service.get()
+        self._course_service = course_service
+
+    def upsert(self, grouping):
+        db_grouping = (
+            self._db.query(Grouping)
+            .filter_by(
+                application_instance=grouping.application_instance,
+                authority_provided_id=grouping.authority_provided_id,
+                type=grouping.type,
+            )
+            .one_or_none()
+        )
+        if not db_grouping:
+            self._db.add(grouping)
+        else:
+            # Update any fields that might have changed
+            db_grouping.lms_name = grouping.name
+            db_grouping.extra = grouping.extra
+
+        return db_grouping or grouping
+
+    def section_grouping(
+        self, tool_consumer_instance_guid, context_id, section_id, section_name
+    ):
+        """
+        Create an HGroup for a course section.
+
+        :param section_name: The name of the section
+        :param tool_consumer_instance_guid: Tool consumer GUID
+        :param context_id: Course id the section is a part of
+        :param section_id: A section id for a section group
+        """
+
+        section_authority_provided_id = hashed_id(
```

Comment:
This needs more information to be unique

will something like `hashed_id(tool_consumer_instance_guid, context_id, "canvas_section", section_id)` work?